### DISCUSSION TITLE
[15488] question on unread_count.

### DISCUSSION THREAD

**Discussion by hehenihao** [eProsima/Fast-DDS#2939] [Q&A] [answered]
I have a BEST_EFFORT publisher which publish a topic every 20ms，and a BEST_EFFORT subscriber which process the message every 25ms. both of them have a sufficient history(255) and KEEP_LAST qos. 

```
        <topic profile_name="topic_profile">
            <kind>NO_KEY</kind>
            <historyQos>
                <kind>KEEP_LAST</kind>
                <depth>255</depth>
            </historyQos>
            <resourceLimitsQos>
                <max_samples>5000</max_samples>
                <max_instances>10</max_instances>
                <max_samples_per_instance>300</max_samples_per_instance>
                <allocated_samples>100</allocated_samples>
	        <extra_samples>20</extra_samples>
            </resourceLimitsQos>
        </topic>
```
In my opinion, I would get 2 unread messages from the reader’s history about each 4 or 5 rounds on subscriber side. But when I print the unread_count, I found it always equal 1.

![image](https://user-images.githubusercontent.com/7553566/187826376-4c7e5cbd-2060-41dd-9e38-b01e7cb1dd93.png)

In my case, i would like to process the unread sample(s) by `on_data_available` notify the consumer. but the consumer may be block by callback in some cases. I treat middleware have already stored all the unhandled messages for me, so I can get and process the message in a loop even if the consumer has been blocked by the former callback.
```
while (GetUnreadCount() > 0) { // call to reader->get_unread_count()
      SampleInfo sinfo;
      ...
      ReturnCode_t ret_code =TakeSampleWithInfo(message.get(), &sinfo); // call to take_next_sample
      if (ret_code == ReturnCode_t::RETCODE_OK) {
        // Trace point of callback execution time
        Callback(message.get());
      }
      ...
    }
```

My question is why I get the unread count always equals 1? Is there a misuse or a misunderstanding of the interfaces?

I'm using fastdds 2.4.1.

**Comment by MiguelCompany** [comment, marked as answer]
@hehenihao I would say that the History QoS of the DataReader is being configured as KEEP_LAST with a depth of 1.

You could check the current value with `data_reader->get_qos().history().depth`.

How are you trying to load the profile you mention?

**Reply by hehenihao** [reply]
@MiguelCompany 
I check the profile of `data_reader`, you are absolutely right. I'm so sorry, but I even had no idea about that before, and I consider the topic history qos is the same as the data_reader's.

I  set the profile like this
```xml
<data_reader profile_name="data_reader_profile">
    <historyMemoryPolicy>PREALLOCATED</historyMemoryPolicy>
    <qos> <!-- writerQosPoliciesType -->
        <durability>
            <kind>VOLATILE</kind>
        </durability>
        <reliability>
            <kind>BEST_EFFORT</kind>
        </reliability>
        <data_sharing>
            <kind>AUTOMATIC</kind>
            <domain_ids>
                <domainId>1</domainId>
            </domain_ids>
        </data_sharing>
    </qos>
</data_reader>
<topic profile_name="topic_profile">
      <kind>NO_KEY</kind>
      <historyQos>
          <kind>KEEP_LAST</kind>
          <depth>10</depth>
      </historyQos>
      <resourceLimitsQos>
          <max_samples>200</max_samples>
          <max_instances>20</max_instances>
          <max_samples_per_instance>100</max_samples_per_instance>
          <allocated_samples>50</allocated_samples>
      </resourceLimitsQos>
  </topic>
```
code as following
```c++
   subscriber_ = participant_->create_subscriber(
        SUBSCRIBER_QOS_DEFAULT);  // create the subscriber
    if (subscriber_ == nullptr) {
      return false;
    }

    topic_ = participant_->create_topic_with_profile(
        m_topic_name, type_->getName(), m_topic_profile_);  // m_topic_profile_ is the profile name of topic，here it values 'topic_profile' as mentioned above
    if (topic_ == nullptr) {
      return false;
    }

    reader_ = subscriber_->create_datareader_with_profile(
        topic_, m_data_reader_profile_, &listener_); // m_data_reader_profile_ is the profile name of data_reader, here it values 'data_reader_profile' as mentioned above
```
I used to think the reader just _inherits_ the history qos setting from the topic, the debug result:

![image](https://user-images.githubusercontent.com/7553566/188475920-2c1317be-8799-4f91-94fa-b39a9dbd5ff6.png)

What's wrong with my profile and code or misunderstanding, thank you in advance.

What's the deffirence between  historyQos of a topic and a reader? And I will try to set the historyQos of data_reader seperately.
